torchtune/modules/vision_transformer.py

```diff
@@ -0,0 +1,334 @@
+            >>> # (bsz, len(indices_return_hidden), num_tiles, num_patches_per_tile + CLS token, embed_dim)
+            >>> print(hidden_states.shape)
+            torch.Size([1, 5, 2, 101, 32]
+        """
+        hidden_states = []
+
+        # parse inputs
+        if images.ndim == 4:
+            n_tiles = 1
+            bsz, nch, w, h = images.shape
+            assert (
+                aspect_ratio is None or aspect_ratio.max() == 1
+            ), f"since images.ndim == 4, aspect ratio must be none or 1x1, got {aspect_ratio}"
+
+        elif images.ndim == 5:
+            bsz, n_tiles, nch, w, h = images.shape
+
+        else:
+            raise ValueError(
+                f"Unsupported number of dimensions: {images.ndim}. Expected 4 or 5."
+            )
+
+        # if aspect_ratio is not provided, it defaults to one tile [1,1]
+        if aspect_ratio is None:
+            aspect_ratio = torch.tensor([[1, 1]] * bsz, device=images.device)
+
+        images = images.reshape(bsz * n_tiles, nch, w, h)
+        aspect_ratio = aspect_ratio.reshape(bsz, 2)
+
+        # patch embeddings (tokens)
+        # out: (bsz, embed_dim, patch_grid_size, patch_grid_size)
+        x = self.conv(images)
+
+        # out: (bsz, patch_grid_size**2, embed_dim)
+        x = x.flatten(start_dim=2).permute(0, 2, 1)
+
+        _, n_tokens, embed_dim = x.shape
+
+        # pre_tile_pos_embed
+        if self.pre_tile_pos_embed:
+            x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+            x = self.pre_tile_pos_embed(x, aspect_ratio)
+
+        # apply cls token
+        x = x.reshape(bsz * n_tiles, n_tokens, embed_dim)
+        x = self.cls_token_embedding(x)
+        n_tokens += 1
+
+        # token_pos_embedding
+        x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+        x = self.token_pos_embedding(x, aspect_ratio)
+
+        # norm
+        x = self.ln_pre(x)
+
+        # transformer with optional hidden layer outputs
+        x = x.view(bsz, n_tiles * n_tokens, embed_dim)
+        for layer_idx, transformer_layer in enumerate(self.transformer_layers):
+            if self.indices_return_hidden and layer_idx in self.indices_return_hidden:
+                hidden_states.append(x)
+            x = transformer_layer(x)
+
+        # norm
+        x = self.ln_post(x)
+
+        # reshape outputs
+        x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+
+        if self.indices_return_hidden:
+            hidden_states = torch.stack(hidden_states, dim=-1)
+            hidden_states = hidden_states.reshape(
+                bsz, n_tiles, n_tokens, embed_dim, len(self.indices_return_hidden)
+            )
+        else:
+            hidden_states = torch.empty(0)  # dummy tensor
+
+        # post_tile_pos_embed
+        if self.post_tile_pos_embed:
+            x = self.post_tile_pos_embed(x, aspect_ratio)
+
+        # cls token projection
+        if self.cls_projection:
+            x = x.reshape(bsz * n_tiles, n_tokens, embed_dim)
```

Comment:
We can do it for modules that contract the shape, but not for those that expand, because it wouldnt know how to from a,b,c to x,y,z,w.

expand: 
```
# from (bsz, patch_grid_size**2, embed_dim)
if self.pre_tile_pos_embed:
    x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
    x = self.pre_tile_pos_embed(x, aspect_ratio)
```

```
# from (bsz * n_tiles, n_tokens, embed_dim)
x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
x = self.token_pos_embedding(x, aspect_ratio)
```

contract:
```
if self.cls_projection:
    x = x.reshape(bsz * n_tiles, n_tokens, embed_dim)
    x = self.cls_projection(x)
```

```
# insert cls token
x = x.reshape(bsz * n_tiles, n_tokens, embed_dim)
x = self.cls_token_embedding(x)
```

I can make a change for the cls_projection and cls_token_embedding, but then the module now has "n_tiles" in it, which is something that it didnt have to be aware previously. What do you think?